Sources/MongoSwift/ReadWriteConcern.swift

```diff
@@ -114,3 +113,166 @@ extension ReadConcern: Equatable {
+        case number(Int32)
+        /// Indicates a tag for nodes that should acknowledge the write. 
+        case tag(String)
+        /// Specifies that a majoirty of nodes should acknowledge the write.
+        case majority
+
+        public init(from decoder: Decoder) throws {
+            let container = try decoder.singleValueContainer()
+            if let wTag = try? container.decode(String.self) {
+                self = .tag(wTag)
+            } else {
+                let wNumber = try container.decode(Int32.self)
+                if wNumber == MONGOC_WRITE_CONCERN_W_MAJORITY {
+                    self = .majority
+                } else {
+                    self = .number(wNumber)
+                }
+            }
+        }
+
+        public func encode(to encoder: Encoder) throws {
+            var container = encoder.singleValueContainer()
+            switch self {
+            case let .number(wNumber):
+                try container.encode(wNumber)
+            case let .tag(wTag):
+                try container.encode(wTag)
+            case .majority:
+                try container.encode("majority")
+            }
+        }
+
+        public static func == (lhs: W, rhs: W) -> Bool {
+            switch (lhs, rhs) {
+            case let (.number(lNum), .number(rNum)):
+                return lNum == rNum
+            case let (.tag(lTag), .tag(rTag)):
+                return lTag == rTag
+            case (.majority, .majority):
+                return true
+            default:
+                return false
+            }
+        }
+    }
+
+    /// Indicates the `W` value for this `WriteConcern`.
+    public var w: W {
+        if let wTag = mongoc_write_concern_get_wtag(self._writeConcern) {
+            return .tag(String(cString: wTag))
+        }
+        let number = mongoc_write_concern_get_w(self._writeConcern)
+        if number == MONGOC_WRITE_CONCERN_W_MAJORITY { return .majority }
+        return .number(number)
+    }
+
+    /// Indicates whether to wait for the write operation to get committed to the journal.
+    public var journal: Bool? {
+        return mongoc_write_concern_get_journal(self._writeConcern)
+    }
+
+    /// If the write concern is not satisfied within this timeout (in milliseconds),
+    /// the operation will return an error. The value MUST be greater than or equal to 0.
+    public var wtimeoutMS: Int32? {
+        return mongoc_write_concern_get_wtimeout(self._writeConcern)
+    }
+
+    /// Indicates whether vbnm this is an acknowledged write concern.
+    public var isAcknowledged: Bool {
+        return mongoc_write_concern_is_acknowledged(self._writeConcern)
+    }
+
+    /// Indicates whether this is the default write concern.
+    public var isDefault: Bool {
+        return mongoc_write_concern_is_default(self._writeConcern)
+    }
+
+    /// Indicates whether the combination of values set on this `WriteConcern` is valid.
+    public var isValid: Bool {
```

Comment:
Is there a need to expose this? PHPC doesn't have a use for it. I realize it's public, but AFAIK libmongoc uses it internally to sanity check a write concern before sending it in a server command. Are there plans to do so in Swift when you manually build a command?

Reply:
Ah, I didn't consider that it might have been used for spec tests. PHPC relies on libmongoc for those, although we do have our own tests for errors thrown during construction of an invalid WriteConcern (e.g. out-of-range `w` value).

I noticed that `WriteConcern.init` (with the `journal`, `w`, and `wtimeoutMS` args) doesn't raise an error for invalid values. Would that make more sense?

Reply:
I considered doing that, but then anytime a user creates a `WriteConcern` they need to use `try`:
```swift
let wc = try WriteConcern(w: 1)
```
which seems kind of annoying, and is asymmetric to how `ReadConcern`s are created. I seem to vaguely recall that in a past version of this PR I was originally doing this and @mbroadst wanted to avoid it. but that was quite a while ago, so I can't say for sure. 

checking validity at the time of encoding (rather than creation time) would prevent us from making the switch to a preconditionFailure in the `description` property though.... hmm.

Reply:
My general feeling is that this is a value class and throwing a `try` in there feels unintuitive. We have a few options then:
 - only allow users to use predefined values, so there are no invalid values
 - use preconditions in the initializer, and hard fail the app when someone screws this up
 - throw from the initializer.

I care about this like 3/10, but not having a throwing initializer "felt" more natural to me